.grit/patterns/hathora_to_new_ts_sdk.md

```diff
@@ -0,0 +1,226 @@
+      $use_args = join(list=[$req, $app_id, $opts], separator=", ")
+    } else if ($args <: [$app_id, $req]) {
+      $use_args = join(list=[$req, $app_id], separator=", ")
+    } else {
+      $use_args = $args
+    }
+  } else {
+    $use_args = join(list=$args, separator=", ")
+  },
+  return $use_args
+}
+
+pattern rewrite_method_calls($resource, $callee) {
+  maybe any {
+    // TODO more elegant handling of await-or-not, maybe via regex? better idea?
+    contains bubble($resource, $callee) `await $callee.$method($args)` as $body where {
+      $suffix = deprecated_suffix(resource=$resource, method=$method),
+      // TODO finish rearranging methods
+      // TODO generalize arg-rearrangement for the non-await case, since they're the same.
+      $use_args = reorder_args($method, $args),
+      $unwrap_at = wrapper_key(method=$method),
+      $body => `(await $callee.$[method]$[suffix]($use_args)).$unwrap_at`
+    },
+    contains bubble($resource, $callee) `$callee.$method($args)` as $body where {
+      $suffix = deprecated_suffix(resource=$resource, method=$method),
+      $use_args = reorder_args($method, $args),
+      $body => `$callee.$[method]$[suffix]($use_args)`
+    }
+  }
+}
+
+pattern generic_replace_resource($resource) {
+  and {
+      replace_resource_import($resource),
+    // TODO what do we do about hygiene when we import names like `sdk`?
+    // do we have to worry about shadowing?
+      maybe any {
+        $member_name = sdk_member_name($resource),      
+        and {
+          contains `const $binding = new $resource()` => `const $binding = sdk.$member_name;`, 
+          rewrite_method_calls(resource=$resource, callee=$binding),
+        },
+        contains type_annotation(type=$resource) where {
+          $source = resource_type_import_source($resource),
+          $substitute = resource_type_substitute($resource),
+          $substitute <: ensure_import_from(source=$source)
+      } => `: $substitute`
+    }
+  }
+}
+
+// v1 functions return an extra layer of indirection now, so eg a login response is at `.loginResponse`
+pattern replace_inside_functions($client_type) {
+  function_declaration(parameters=$params, body=$body) where {
+    // TODO critically, this depends on types. if the target project is pure js, we're SoL.
+    // One option is to just assume any name calling a given method like `getConnectionInfo` is using Hathora.
+    // Not every method name may be that distinctive, and in any real codebase it could be insanely noisy.
+    // A more invasive solution that seems like it would require Grit core support
+    // is just following variable refs by substitution (where possible!) and guessing what's what.
+    // There is a point at which that's just reproducing a static analyzer.
+    $params <: [..., `$client_binding: $client_type`, ...],
+    $body <: rewrite_method_calls(resource=$client_type, callee=$client_binding),
+  }
+}
+
+pattern gather_calls() {
+  and { 
+    any {
```

Comment:
any any should be redundant - why not just `any` at the root?

Reply:
👍  Definitely just a leftover dirty-code issue. TODO: write a `prettier` for Grit.

Reply:
That's in our backlog. :)